### DISCUSSION TITLE
Reseting/setting new query without breaking the QB

### DISCUSSION THREAD

**Discussion by giero** [react-querybuilder/react-querybuilder#794] [Q&A] [answered]
Hello,

I would like to set the query from the outside - in my case I have a button which sets query to the default state ... however this approach breaks the QB - after setting new query value editor loses focus after each keystroke.

You can reproduce this in my [sandbox](https://codesandbox.io/p/devbox/wispy-sound-2q2cgp?file=%2Fsrc%2FApp.tsx) - and now I don't know if it's a bug or a feature because I should have done it differently (unfortunately I didn't find the answer anywhere how to reset the query).

**Comment by jakeboone02** [comment, marked as answer]
I'm not sure whether to call it a bug or a feature either, to be honest! It has to do with a performance optimization that's based on an assumption that fails in your case. The relevant code, with a comment addressing the assumption, is here:

https://github.com/react-querybuilder/react-querybuilder/blob/6b6e9b71b73186a405f325ebb61f0cb45e9abb2b/packages/react-querybuilder/src/hooks/useQueryBuilderSchema.ts#L191-L197

The optimization is to skip the `prepareRuleGroup` call if it thinks the `query` is already "prepared" (i.e., each rule and subgroup has a unique `id`). Without an `id` property on every object, you get the "lose focus" behavior you're seeing.

I'm not sure if I'll do anything about it, but in the meantime you've got several options for workarounds:

1. Seed `initialQuery` with `id`s by running it through `prepareRuleGroup` in the declaration. (You could also declare the `id` properties manually, just make sure you cover _all_ rules and subgroups.)
    ```ts
    const initialQuery: RuleGroupType = prepareRuleGroup({
      combinator: 'and',
      rules: [
        { field: 'firstName', operator: 'beginsWith', value: 'Stev' },
        { field: 'lastName', operator: 'in', value: 'Vai,Vaughan' },
      ],
    });
    ```
2. Run `prepareRuleGroup` before resetting the state:
    ```diff
    -      <button id="reset" onClick={() => setQuery(initialQuery)}>
    +      <button id="reset" onClick={() => setQuery(prepareRuleGroup(initialQuery))}>
    ```
3. Use a `key` prop and change it when you want to force React to render the component from scratch:
    ```ts
    export const App = () => {
      const [k, setK] = useState('initial');
      const [query, setQuery] = useState(initialQuery);
    
      return (
        <div>
          <button
            id="reset"
            onClick={() => {
              setK(crypto.randomUUID());
              setQuery(initialQuery);
            }}>
            RESET QB
          </button>
          <QueryBuilder
            key={k}
            fields={fields}
            query={query}
            onQueryChange={setQuery}
          />
        </div>
      );
    };
    ```

Hope that helps!

**Reply by giero** [reply]
Works like a charm :) Thank you very much for fast response :+1: 

`prepareRuleGroup` and defining `key` was what I needed.
